runtime/jailer.go

```diff
@@ -0,0 +1,403 @@
+	rootPath := j.RootPath()
+	*socketPath = filepath.Join(rootPath, "api.socket")
+
+	return firecracker.Handler{
+		Name: jailerHandlerName,
+		Fn: func(ctx context.Context, m *firecracker.Machine) error {
+			const mode = os.FileMode(0700)
+			// Create the proper paths needed for the runc jailer
+			j.logger.WithField("rootPath", rootPath).Debug("Creating root drive path")
+			if err := os.MkdirAll(rootPath, mode); err != nil {
+				return errors.Wrapf(err, "failed to create root path: %v", rootPath)
+			}
+
+			binPath := filepath.Join(rootPath, "usr", "local", "bin")
+			j.logger.WithField("binPath", binPath).Debug("Creating /usr/local/bin")
+			if err := os.MkdirAll(binPath, mode); err != nil {
+				return errors.Wrapf(err, "failed to create /usr/local/bin in root path: %v", binPath)
+			}
+
+			contentsPath := j.ContentsPath()
+			j.logger.WithField("contentsPath", contentsPath).Debug("Creating firecracker contents path")
+			if err := os.MkdirAll(contentsPath, mode); err != nil {
+				return errors.Wrapf(err, "failed to create contents path: %v", contentsPath)
+			}
+
+			rootPathToConfig := filepath.Join(ociBundlePath, "config.json")
+			j.logger.WithField("rootPathToConfig", rootPathToConfig).Debug("Copying config")
+			if err := copyFile(runcConfigPath, rootPathToConfig, 0666); err != nil {
+				return errors.Wrapf(err, "failed to copy config from %v to %v", runcConfigPath, rootPathToConfig)
+			}
+
+			j.logger.Debug("Overwritting process args of config")
+			if err := j.overwriteConfig(cfg, filepath.Base(m.Cfg.SocketPath), rootPathToConfig); err != nil {
+				return errors.Wrap(err, "failed to overwrite config.json")
+			}
+
+			// copy the firecracker binary
+			j.logger.WithField("binPath", binPath).Debug("copying firecracker binary")
+			if err := copyFile(cfg.FirecrackerBinaryPath, binPath, 0700); err != nil {
+				return errors.Wrapf(err, "could not copy firecracker binary from path %v", cfg.FirecrackerBinaryPath)
+			}
+
+			// Link the kernel image
+			newKernelImagePath := filepath.Join(contentsPath, kernelImageFileName)
+			j.logger.WithField("newKernelImagePath", newKernelImagePath).Debug("linking kernel image")
+			if err := os.Link(m.Cfg.KernelImagePath, newKernelImagePath); err != nil {
+				return errors.Wrapf(err, "failed to link kernel image from path %v", m.Cfg.KernelImagePath)
+			}
+			m.Cfg.KernelImagePath = filepath.Join("/var/lib/firecracker-containerd", kernelImageFileName)
+
+			// link drives to new contents path
+			for i, d := range m.Cfg.Drives {
+				drivePath := firecracker.StringValue(d.PathOnHost)
+				fileName := filepath.Base(drivePath)
+
+				if err := os.Link(drivePath, filepath.Join(contentsPath, fileName)); err != nil {
+					return errors.Wrapf(err, "failed to copy drive %v", drivePath)
+				}
+
+				newDrivePath := filepath.Join(strings.TrimPrefix(contentsPath, rootPath), fileName)
+				m.Cfg.Drives[i].PathOnHost = firecracker.String(newDrivePath)
+			}
+
+			if err := chownr(
```

Comment:
I'm concerned about this. We're manipulating metadata associated with files that are hardlinked into the jail. Thus, we're making changes outside the jail. Even if we do want the files to be owned by the jailer UID, we shouldn't be setting that here. Further, when we're using a unique UID/GID per VM, this is going to cause problems.

We really shouldn't be doing a recursive chown across the jailer directory. We know exactly what files we're installing, and why we're installing them, so we should install them with appropriate mode bits from the outset. Files that are shared between VMs should be owned by root and shared with minimal permission into the VM (read-only, so probably mode 444). Files that are not shared can get the appropriate metadata settings when we copy them into place.

Reply:
How about chown everything except the hard linked files? Taking a look at Firecracker's jailer this is what they do. For the hard linked files, I am unsure how we want to handle this since the drives need to be r/w, so doing read only won't work. However, for the kernel image this makes sense. I am pretty sure we don't want to copy the drive since that can be pretty large, so the only option I can think of is chown. That's what I did when running with the Firecracker's jailer, I'd chown the drive. What do you prefer we do with drives?

Reply:
Drives for e.g. the VM root filesystem _must_ be shared read-only into the jail. Note that the root filesystem generated by `image-builder` is a squashfs, which doesn't even support write operations anyway.

Only the devices backing the container storage should be read/write in the jail, and they should be exposed as block-devices by `mknod()` rather than being bind-mounted. In those cases, you're already setting the appropriate access metadata in `ExposeDeviceToJail()`, so the `chownr()` call shouldn't be necessary.

Reply:
Ah! Good point. Okay, will go ahead and make the changes. Thank you